Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/ThresholdedRandomCutForest.java

```diff
@@ -432,6 +442,134 @@ public void setInitialThreshold(double initial) {
+                /**
+                 * adds information of expected point to the result descriptor (provided it is
+                 * marked anomalous) Note that is uses relativeIndex; that is, it can determine
+                 * that the anomaly occurred in the past (but within the shingle) and not at the
+                 * current point -- even though the detection has triggered now While this may
+                 * appear to be improper, information theoretically we may have a situation
+                 * where an anomaly is only discoverable after the "horse has bolted" -- suppose
+                 * that we see a random mixture of the triples { 1, 2, 3} and {2, 4, 5}
+                 * corresponding to "slow weeks" and "busy weeks". For example 1, 2, 3, 1, 2, 3,
+                 * 2, 4, 5, 1, 2, 3, 2, 4, 5, ... etc. If we see { 2, 2, X } (at positions 0 and
+                 * 1 (mod 3)) and are yet to see X, then we can infer that the pattern is
+                 * anomalous -- but we cannot determine which of the 2's are to blame. If it
+                 * were the first 2, then the detection is late. If X = 3 then we know it is the
+                 * first 2 in that unfinished triple; and if X = 5 then it is the second 2. In a
+                 * sense we are only truly wiser once the bolted horse has returned! But if we
+                 * were to say that the anomaly was always at the second 2 then that appears to
+                 * be suboptimal -- one natural path can be based on the ratio of the triples {
+                 * 1, 2, 3} and {2, 4, 5} seen before. Even better, we can attempt to estimate a
+                 * dynamic time dependent ratio -- and that is what RCF would do.
+                 *
+                 * @param result the description of the current point
+                 */
+                int shingleSize = result.getShingleSize();
+                int dimension = result.getDimension();
+                int base = dimension / shingleSize;
+                double[] reference = result.getCurrentInput();
+                float[] newPoint = result.getExpectedRCFPoint();
+
+                int index = result.getRelativeIndex();
+                if (index < 0) {
+                    reference = preprocessor.getShingledInput(shingleSize + index);
+                    result.setPastTimeStamp(preprocessor.getTimeStamp(shingleSize + index));
+                }
+                result.setPastValues(reference);
+                if (newPoint != null) {
+                    double[] values = preprocessor.getExpectedValue(index, reference, point, newPoint);
+                    if (forestMode == ForestMode.TIME_AUGMENTED) {
+                        int endPosition = (shingleSize - 1 + index + 1) * dimension / shingleSize;
+                        double timeGap = (newPoint[endPosition - 1] - point[endPosition - 1]);
+                        long expectedTimestamp = (timeGap == 0) ? result.getInputTimestamp()
```

Comment:
No No -- the time gap here corresponds to the question "was the anomaly (likely) caused by a difference in the input timestamp which is in coordinate (end position - 1)". If time gap == 0, then that means that the input timestamp was not a factor and will use the input timestamp for the expected timestamp. If the gap is not 0, then we have to compute inverses etc. The reason for doing this was that we use single precision float internally, so if we always use the inverses then we will have 0.9999, 2.000001 like values instead of 1, 2 etc... That can be very annoying (and can cause more trouble). In fact this is the same reason (not to have floating point approximations all over the place) I decided to keep the inputs double precision in several places. The approximations still show up sometimes -- but are few and far between.

Reply:
Not sure if I understand. I will ask offline.